### PR TITLE
Feature/metobs

### PR DESCRIPTION
**Pull request by mgcth**
Clean MetObs module and add tests for it.

### CODE REVIEW

**Review by docNord — COMMENTED** [review]
When calling inspect(), the first number presented is an irrelevant index which can throw a user (such as me) off. I suggest removing it or making it less pronounced in the presentation of information.

**Comment by mgcth** [comment]
Fixed in [bdf5a3d](https://github.com/Ingenjorsarbete-For-Klimatet/smhi/pull/4/commits/bdf5a3df6514c8bea2e18efe3e53798b029bcee7).

**Review by docNord — COMMENTED** [review]
It seems reasonable to use only "get" in place of the sometimes used "fetch", as get is more in line with the http protocol.

**Comment by mgcth** [comment]
> It seems reasonable to use only "get" in place of the sometimes used "fetch", as get is more in line with the http protocol.

Agree, but try to use the review function here to point where in the code :).

File: smhi/metobs.py [hunk 1]
```diff
@@ -36,76 +41,109 @@ def __init__(self, type: str = "json"):
         self.table_raw = None
         self.table = None
 
-    def fetch_parameters(self, version: str = "1.0"):
+    def fetch_parameters(self, version: Union[str, int] = "1.0"):
```
Comment (docNord) [inline smhi/metobs.py:44]: def get_parameters(self, version: Union[str, int] = "1.0");

File: smhi/metobs.py [hunk 2]
```diff
@@ -36,76 +41,109 @@ def __init__(self, type: str = "json"):
...
-            raise Exception("Only supports version = 1.0.")
+            raise NotImplementedError(
+                "Version {} not supported. Only supports version = 1.0.".format(version)
+            )
 
         self.version = version
-        self.parameter = SMHIParameterV1(self.type, self.available_versions, version)
+        self.parameter = MetObsParameterV1(self.available_versions)
 
     def fetch_stations(self, parameter: str = None, parameter_title: str = None):
```
Comment (docNord) [inline smhi/metobs.py:62]: def get_stations(self, parameter: str = None, parameter_title: str=None):

File: smhi/metobs.py [hunk 3]
```diff
@@ -36,76 +41,109 @@ def __init__(self, type: str = "json"):
...
         """
-        Fetch SMHI API (version 1) stations from given parameter.
+        Fetch SMHI MetObs API (version 1) stations from given parameter.
```
Comment (docNord) [inline smhi/metobs.py:64]: ... in the spirit of consistency might as well change from Fetch to Get here too.

File: smhi/metobs.py [hunk 4]
```diff
@@ -36,76 +41,109 @@ def __init__(self, type: str = "json"):
...
-                self.type, self.parameter.resource, parameter_title=parameter_title
+        if parameter:
+            self.station = MetObsStationV1(self.parameter.resource, parameter=parameter)
+        else:
+            self.station = MetObsStationV1(
+                self.parameter.resource, parameter_title=parameter_title
             )
 
-    def fetch_periods(self, station: str = None, station_set: str = None):
+    def fetch_periods(self, station: str = None, stationset: str = None):
```
Comment (docNord) [inline smhi/metobs.py:80]: def get_periods(self, station: str = None, stationset: str = None):

File: smhi/metobs.py [hunk 5]
```diff
@@ -36,76 +41,109 @@ def __init__(self, type: str = "json"):
...
-        if station is None:
-            self.period = SMHIPeriodV1(
-                self.type, self.station.station, station_set=station_set
-            )
+        if station:
+            self.period = MetObsPeriodV1(self.station.station, station)
+        else:
+            self.period = MetObsPeriodV1(self.station.station, stationset=stationset)
 
     def fetch_data(self, period: str = "corrected-archive"):
```
Comment (docNord) [inline smhi/metobs.py:96]: def get_data(self, period: str = "corrected-archive"):

File: smhi/metobs.py [hunk 6]
```diff
@@ -36,76 +41,109 @@ def __init__(self, type: str = "json"):
...
         """
-        Fetch SMHI API (version 1) data from given period.
+        Fetch SMHI MetObs API (version 1) data from given period.
```
Comment (docNord) [inline smhi/metobs.py:98]: Get

File: smhi/metobs.py (hunk 5 above)
Comment (docNord) [reply smhi/metobs.py:96]: There is another "get_data" below, hence the double naming must be resolved before removing the "fetch" terminology

File: smhi/metobs.py [hunk 7]
```diff
@@ -36,76 +41,109 @@ def __init__(self, type: str = "json"):
...
 
         Args:
             period: period
         """
-        self.data = SMHIDataV1(self.type, self.period.period, period)
+        self.data = MetObsDataV1(self.period.period, period)
         self.table_raw = self.data.fetch()
         self.table = pd.read_csv(io.StringIO(self.table_raw), on_bad_lines="skip")
 
+    def get_data(
```
Comment (docNord) [inline smhi/metobs.py:107]: rename?

File: smhi/metobs.py (hunk 4 above)
Comment (mgcth) [reply smhi/metobs.py]: Fixed in e2356c1.

File: smhi/metobs.py (hunk 7 above)
Comment (mgcth) [reply smhi/metobs.py]: Renamed to `get_from_selection` in e2356c1. Is `get` enought?

File: smhi/metobs.py [hunk 8]
```diff
@@ -36,76 +41,109 @@ def __init__(self, type: str = "json"):
...
+        else:
+            self.station = MetObsStationV1(
+                self.parameter.resource, parameter_title=parameter_title
             )
 
-    def fetch_periods(self, station: str = None, station_set: str = None):
+    def fetch_periods(self, station: str = None, stationset: str = None):
         """
-        Fetch SMHI API (version 1) periods from given station.
+        Fetch SMHI MetObs API (version 1) periods from given station.
```
[the remark this answers is not in the document]
Comment (mgcth) [reply smhi/metobs.py]: Fixed in 343fd56.

File: smhi/metobs.py [hunk 9]
```diff
@@ -102,7 +102,9 @@ def get_data(self, period: str = "corrected-archive"):
         """
         self.data = MetObsDataV1(self.period.period, period)
         self.table_raw = self.data.get()
-        self.table = pd.read_csv(io.StringIO(self.table_raw), on_bad_lines="skip")
+        data_starting_point=self.table_raw.find('Datum')
+        self.header=self.table_raw[0:data_starting_point]
+        self.table = pd.read_csv(io.StringIO(self.table_raw[data_starting_point:-1]),sep=';', on_bad_lines="skip",usecols=[0,1,2])
```
Comment (mgcth) [inline smhi/metobs.py:107]: Looks good. How is no data handled? I get back a table with a few rows of NaNs now.

Also, try using black (do `pip install black` and activate `Save on format` in VSCode). It'll format your code for you every time you save, i.e, space between `=` etc. I'll add a format check later as a GitHub Action. You could also try pre-commit (also `pip install pre-commit`), which does some checks every time you commit. We have a `.pre-commit-config.yaml` file in the repo root which should be used if you do `pre-commit install`. This won't let you commit until some checks are fulfilled :).

Comment (mgcth) [reply smhi/metobs.py:107]: Also, I see the intergation test fails. Try to update the test when you make changes to the code. To run a test locally it should be enough to write `pytest` if it is installed (also with pip).

Comment (docNord) [reply smhi/metobs.py:107]: Thanks - will update tomorrow. Have installed the suggested packages, will look into VSCode settings tomorrow with the work.

**Comment by docNord** [comment]
I have installed black as requested above, modified the test (not very comprehensive, but works). When there is no data, the table is filled with NaNs, which I think is alright all things considered. Data is parsed to the extent that it is directly useable, although I have not bothered to change the dates to datetime (it is one line of code, doesn't seem necessary).

File: smhi/metobs.py (hunk 5 above)
Comment (docNord) [reply smhi/metobs.py]: Enough.

**Comment by docNord** [comment]
@mgcth I am unable to add you as reviewer (possibly since you are the assignee - I changed that so that I was the assignee instead, but that did not seem to address the issue. Can you make yourself reviewer, you think?

**Comment by mgcth** [comment]
> @mgcth I am unable to add you as reviewer (possibly since you are the assignee - I changed that so that I was the assignee instead, but that did not seem to address the issue. Can you make yourself reviewer, you think?

No, probably because I've written most of the code. But I can review your changes just fine without being assigned.

File: .vscode/settings.json [hunk 10]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (mgcth) [inline .vscode/settings.json:1]: Should probably not commit this file.

File: smhi/metobs.py [hunk 11]
```diff
@@ -102,9 +102,14 @@ def get_data(self, period: str = "corrected-archive"):
         """
         self.data = MetObsDataV1(self.period.period, period)
         self.table_raw = self.data.get()
-        data_starting_point=self.table_raw.find('Datum')
-        self.header=self.table_raw[0:data_starting_point]
-        self.table = pd.read_csv(io.StringIO(self.table_raw[data_starting_point:-1]),sep=';', on_bad_lines="skip",usecols=[0,1,2])
+        data_starting_point = self.table_raw.find("Datum")
+        self.header = self.table_raw[0:data_starting_point]
+        self.table = pd.read_csv(
```
Comment (mgcth) [inline smhi/metobs.py:107]: Would probably make more sense to return both header and data and not store them inside the object.

**Review by mgcth — COMMENTED** [review]
Looks good, just two comments.

File: .vscode/settings.json (hunk 10 above)
Comment (docNord) [reply .vscode/settings.json:1]: Sorry. I suppose I ought to add that to the .gitignore then?

Comment (mgcth) [reply .vscode/settings.json:1]: Yes. Honestly, I thought it was included already.

Comment (mgcth) [reply .vscode/settings.json:1]: That is, add the whole folder `.vscode`.

**Comment by docNord** [comment]
I am calling it a night now @mgcth - sorted the integration tests out to some degree, will look into the unit tests tomorrow. Some things changed when restructuring, obviously.

**Comment by mgcth** [comment]
> I am calling it a night now @mgcth - sorted the integration tests out to some degree, will look into the unit tests tomorrow. Some things changed when restructuring, obviously.

Great! Yeah, sorry for dragging main in here a bit too quick. If we merge #17 first, I can try to sort the conflicts out here.

File: test/unit/test_unit_metobs.py [hunk 12]
```diff
@@ -0,0 +1,691 @@
...
+        assert client.table_raw == mock_metobsdatav1.return_value.get()
+
+        mock_metobsdatav1.assert_called_once()
+        mock_pd_read_csv.assert_called_once()
+        mock_stringio.assert_called_once()
+
+    @patch("smhi.metobs.MetObs.get_parameters")
+    @patch("smhi.metobs.MetObs.get_stations")
+    @patch("smhi.metobs.MetObs.get_periods")
+    @patch("smhi.metobs.MetObs.get_data")
```
Comment (mgcth) [inline test/unit/test_unit_metobs.py:199]: Here, write something like
`@patch("smhi.metobs.MetObs.get_data", return_value=("test1", "test2"))`
OR
before line 217 do
`mock_get_data.return_value = ("test1, "test2")`

and check that `client.get_data_from_selection` actually returns that.

**Review by docNord — DISMISSED** [review]
All changes are mine, but assuming you don't have further comments @mgcth ...

**Comment by mgcth** [comment]
@docNord, we need to merge main into this, I can do that.

**Comment by mgcth** [comment]
But can you first approve #17 so we get those changes in here.

**Comment by mgcth** [comment]
@docNord, before merging, we should probably add some example code to the docs. Do you want to do that? I can clean up the markdown later (and will add a check for that too). In the meantime, you can maybe use the VSCode extension *Markdown All in One* or something?

File: test/unit/test_unit_metobs.py (hunk 12 above)
Comment (mgcth) [reply test/unit/test_unit_metobs.py]: Resolved in [ea44e0c](https://github.com/Ingenjorsarbete-For-Klimatet/ifk-smhi/pull/4/commits/ea44e0c76c9cdeaec92fde9a0d9b5903d213f9d7).

File: smhi/metobs.py (hunk 11 above)
Comment (mgcth) [reply smhi/metobs.py]: Fixed in [873649a](https://github.com/Ingenjorsarbete-For-Klimatet/ifk-smhi/pull/4/commits/873649a3a98464167adb88d92667208d67a82926).

**Comment by docNord** [comment]
@mgcth : I added the Markdown package to my vscode installation and included a simple example similar to your example from Strang. See what you think of it.

**Comment by docNord** [comment]
I see that you return a "status" variable from Strang. I have never been a huge fan, but perhaps I should add that to the MetObs setup so that we use the same return policy? Also, should I spend time to parse the header? (please say no ;) )

**Comment by mgcth** [comment]
> I see that you return a "status" variable from Strang. I have never been a huge fan, but perhaps I should add that to the MetObs setup so that we use the same return policy? Also, should I spend time to parse the header? (please say no ;) )

Shoild probably return similar things. We don't have to return the status.

File: docs/metobs-example.md [hunk 13]
```diff
@@ -0,0 +1,23 @@
+
+Direct usage of `MetObs`, without using the `SMHI` client is possible.
```
Comment (mgcth) [inline docs/metobs-example.md:2]: Could be nice to add examples of usage of the `get_data` method, e.g. get station, parameter call chain etc.

Comment (mgcth) [reply docs/metobs-example.md:3]: So, I implemented this in [63c53b9](https://github.com/Ingenjorsarbete-For-Klimatet/ifk-smhi/pull/4/commits/63c53b9fb6f84bacbdd687ae67bee672d4871204). @docNord, please have a look and see if you like it.

**Comment by mgcth** [comment]
I'm thinking maybe we should rename some attributes

- `parameter`->`parameters`
- `station`->`stations`
- `period`->`periods`

as they hold more than one item (in general).

Further, should we rename the object `MetObs`->`Metobs` maybe?

**Comment by docNord** [comment]
That makes a lot of sense @mgcth . Should I do it? I still have that update of the documentation to sort out, could include it in the same update. Also, should I align the markdown with this new test you have added?

**Comment by mgcth** [comment]
I've fixed the docs, but you can update the code if you want. We can fix the markdown later later. Need to close this PR asap I think :)

File: src/smhi/metobs.py [hunk 14]
```diff
@@ -57,50 +56,54 @@ def get_parameters(self, version: Union[str, int] = "1.0"):
             )
 
         self.version = version
-        self.parameter = MetObsParameterV1(self.available_versions)
+        self.parameters = MetobsParametersV1(self.available_versions)
 
-    def get_stations(self, parameter: str = None, parameter_title: str = None):
+    def get_stations(self, parameters: str = None, parameters_title: str = None):
```
Comment (mgcth) [inline src/smhi/metobs.py:61]: Sorry for being pedantic, but I think for function parameters (and arguments) it makes more sens to keep the singular naming, e.g. parameter, as only one parameter is sent in.

**Comment by docNord** [comment]
I knew you were going to say that! :D Ah well, I'll have one more go.

**Comment by mgcth** [comment]
@docNord, I've fixed it, I think. Have a look and approve if you think we are done here for now.

**Review by docNord — APPROVED** [review]
Nice work!

**Pull request merged**